File: 34-FindFirstAndLastPositionOfElementInSortedArray/34-FindFirstAndLastPositionOfElementInSortedArray.java
// Last updated: 7/28/2025, 12:33:03 PM
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int n=nums.length;
        int lb=lowerbound(nums,n,target);
        if(lb==n || nums[lb]!=target)  return new int[] {-1,-1};
        int ub=upperbound(nums,n,target);
        return new int[] {lb,ub};
    }
   static int lowerbound(int[] nums , int n , int k){
        int low=0,high=n-1 , ans=n, mid=0;
        while(low<=high){
           mid=(low+high)/2;
           if(nums[mid]>=k){
               ans=mid;
               high=mid-1;
           }
           else{
               low=mid+1;
           }

        }
        return ans;
    }
   static int upperbound(int[] nums , int n , int k){
        int low=0,high=n-1 , ans=n, mid=0;
        while(low<=high){
           mid=(low+high)/2;
           if(nums[mid]>k){
               ans=mid;
               high=mid-1;
           }
           else{
               low=mid+1;
           }
        }
        return ans-1;
    }
    };

File: 34-FindFirstAndLastPositionOfElementInSortedArray/34-FindFirstAndLastPositionOfElementInSortedArray.cpp
// Last updated: 7/28/2025, 12:32:56 PM
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int fir = first(nums , target);
        int sec = last(nums,target);
        return {fir , sec};
    }
    int first(vector<int>& nums, int target){
        int low =  0 , high = nums.size() - 1 ,ans = -1;
        while(low <= high){
            int mid = (low+high)/2;
            if(nums[mid]==target){
                ans = mid;
                high = mid -1;
            }
            else if(nums[mid]<target) {
                low = mid+1;
            }
            else {
                high = mid - 1;
            }
        }
        return ans;
    }
       int last(vector<int>& nums, int target){
        int low =  0 , high = nums.size() - 1 ,ans = -1;
        while(low<=high){
            int mid = (low+high)/2;
            if(nums[mid]==target){
                ans = mid;
                low = mid+1;
            }
            else if(nums[mid] < target) {
                low = mid +1;
            }
            else {
                high = mid -1;
            }
        }
        return ans;
    }
};